File: 2022/05/17/intersectionTwoArrays.py
"""
https://leetcode.com/problems/intersection-of-two-arrays-ii/
Author: Andrew Jarombek
Date: 5/17/2022
"""

from typing import List


def intersect(nums1: List[int], nums2: List[int]) -> List[int]:
    elements = {}

    for num in nums1:
        elements[num] = elements.get(num, 0) + 1

    result = []
    for num in nums2:
        val = elements.get(num, 0)

        if val > 0:
            result.append(num)
            elements[num] = val - 1

    return result

File: 2023/06/10/go/intersection_two_arrays.go
/**
 * https://leetcode.com/problems/intersection-of-two-arrays/
 * Author: Andrew Jarombek
 * Date: 6/10/2023
 */

package _go

func intersect(nums1 []int, nums2 []int) []int {
	var chars = make(map[int]int)

	for _, num := range nums1 {
		chars[num] = chars[num] + 1
	}

	var result []int

	for _, num := range nums2 {
		if chars[num] > 0 {
			result = append(result, num)
			chars[num] = chars[num] - 1
		}
	}

	return result
}
